galasa-ui/src/components/runs/LogTab.tsx

```diff
@@ -35,100 +54,288 @@ export default function LogTab({ logs }: { logs: string }) {
     }));
   };
 
+  const toggleMatchCase = () => {
+    setMatchCase(!matchCase);
+  };
+
+  const toggleMatchWholeWord = () => {
+    setMatchWholeWord(!matchWholeWord);
+  };
+
+  const goToNextMatch = () => {
+    if (totalMatches > 0) {
+      setCurrentMatchIndex((prev) => (prev + 1) % totalMatches);
+    }
+  };
+
+  const goToPreviousMatch = () => {
+    if (totalMatches > 0) {
+      setCurrentMatchIndex((prev) => (prev - 1 + totalMatches) % totalMatches);
+    }
+  };
+
+  const createSearchRegex = (term: string) => {
+    if (!term.trim()) return null;
+
+    let escapedTerm = term.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+    if (matchWholeWord) {
+      escapedTerm = `\\b${escapedTerm}\\b`;
+    }
+    const flags = matchCase ? 'g' : 'gi';
+    return new RegExp(escapedTerm, flags);
+  };
+
+  const getLogLevel = (line: string) => {
+    // Check for explicit log levels in the line
+    if (line.includes(' ERROR ')) return 'ERROR';
+    if (line.includes(' WARN ')) return 'WARN';
+    if (line.includes(' DEBUG ')) return 'DEBUG';
+    if (line.includes(' INFO ')) return 'INFO';
+    if (line.includes(' TRACE ')) return 'TRACE';
+
+    // If no explicit level found, return null (will be assigned to previous level)
+    return null;
+  };
+
+  const processLogLines = (content: string) => {
+    const lines = content.split('\n');
+    const processed: LogLine[] = [];
+    let currentLevel = 'INFO'; // Default level
+
+    lines.forEach((line, index) => {
+      const detectedLevel = getLogLevel(line);
+
+      // If we find a new level, update current level
+      if (detectedLevel) {
+        currentLevel = detectedLevel;
+      }
+
+      // All lines get assigned to the current level (either explicit or inherited)
+      processed.push({
+        content: line,
+        level: currentLevel,
+        lineNumber: index + 1,
+        isVisible: true
+      });
+    });
+
+    return processed;
+  };
+
+  const applyFilters = (lines: LogLine[]) => {
+
+    const hasActiveFilters = Object.values(filters).some(filter => filter === true);
+
+    if (!hasActiveFilters) {
+      // none selected → hide all
+      return lines.map(line => ({ ...line, isVisible: false }));
+    }
+
+    // only show lines whose level is still checked
+    return lines.map(line => ({
+      ...line,
+      isVisible: !!filters[line.level as keyof typeof filters]
+    }));
+  };
+
+  const highlightText = (text: string, searchTerm: string, lineIndex: number) => {
+    if (!searchTerm.trim()) return text;
+
+    const regex = createSearchRegex(searchTerm);
+    if (!regex) return text;
+
+    const parts = text.split(regex);
+    const matches = text.match(regex);
+
+    if (!matches) return text;
```

Comment:
Same here regarding lots of return statements, which can make reading through this code more challenging